Make UB transmutes really UB in LLVM

scottmcm opened the pull request:
Ralf suggested in <https://github.com/rust-lang/rust/pull/143410#discussion_r2184928123> that UB transmutes shouldn't be trapping, which happened for the one path *that* PR was changing, but there's another path as well, so *this* PR changes that other path to match.

r? codegen

scottmcm committed:
Make UB transmutes really UB in LLVM

Ralf suggested in <https://github.com/rust-lang/rust/pull/143410#discussion_r2184928123> that UB transmutes shouldn't be trapping, which happened for the one path that PR was changing, but there's another path as well, so this PR changes that other path to match.

RalfJung commented:
@nikic is this the best way to indicate "unreachable" to LLVM in the middle of a block? Does LLVM quickly normalize this by making the entire block `unreachable` or can such an `assume` cause problems?

dianqk commented:
> @nikic is this the best way to indicate "unreachable" to LLVM in the middle of a block? Does LLVM quickly normalize this by making the entire block `unreachable` or can such an `assume` cause problems?

According to https://llvm.org/docs/Frontend/PerformanceTips.html#id11, this should be `store i1 true, ptr poison, align 1`.

RalfJung commented:
Ah! We should have a helper method for this then.

dianqk commented:
> Ah! We should have a helper method for this then.

Yes. This makes sence to me.
BTW, this reminds me of #127740.

nikic commented:
FWIW I don't think it particularly matters how Rust emits this.

scottmcm committed:
Add `BuilderMethods::unreachable_nonterminator`

So places that need `unreachable` but in the middle of a basic block can call that instead of figuring out the best way to do it.

bors commented:
:pushpin: Commit f5fc8727dbbf8c9e93bb0822b2e5bfa77dbd0208 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin f5fc8727dbbf8c9e93bb0822b2e5bfa77dbd0208 -->
<!-- homu: {"type":"Approved","sha":"f5fc8727dbbf8c9e93bb0822b2e5bfa77dbd0208","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
